include/madwf_ml.h

```diff
@@ -0,0 +1,372 @@
+
+    // The parameters to be trained.
+    using device_container = device_vector<transfer_float>;
+    device_container device_param;
+
+    // The diagonal component to suppress the zero modes.
+    double mu;
+    int Ls_base;
+
+    int null_miniter;
+    double null_tol;
+    int train_maxiter;
+
+    // persistent buffers for reuse.
+    ColorSpinorField *forward_tmp = nullptr;
+    ColorSpinorField *backward_tmp = nullptr;
+
+    bool param_load;
+    bool param_save;
+
+    char param_infile[256];
+    char param_outfile[256];
+
+    // Has device_param been trained?
+    bool trained = false;
+
+    QudaPrecision prec_precondition;
+
+    static std::unordered_map<std::string, std::vector<transfer_float>> host_training_param_cache; // empty map
+
+    MadwfAcc(const SolverParam &solve_param) :
+      mu(solve_param.madwf_diagonal_suppressor),
+      Ls_base(solve_param.madwf_ls),
+      null_miniter(solve_param.madwf_null_miniter),
+      null_tol(solve_param.madwf_null_tol),
+      train_maxiter(solve_param.madwf_train_maxiter),
+      param_load(solve_param.madwf_param_load == QUDA_BOOLEAN_TRUE),
+      param_save(solve_param.madwf_param_save == QUDA_BOOLEAN_TRUE),
+      prec_precondition(solve_param.precision_precondition)
+    {
+      strcpy(param_infile, solve_param.madwf_param_infile);
+      strcpy(param_outfile, solve_param.madwf_param_outfile);
+
+      printfQuda("Launching MADWF accelerator ... \n");
+      printfQuda("madwf_mu (low modes suppressor)                   = %.4f\n", mu);
+      printfQuda("madwf_ls (cheap Ls)                               = %d\n", Ls_base);
+      printfQuda("madwf_null_miniter                                = %d\n", null_miniter);
+      printfQuda("madwf_null_tol                                    = %4.2e\n", null_tol);
+      printfQuda("madwf_train_maxiter (max # of iters for training) = %d\n", train_maxiter);
+    }
+
+    ~MadwfAcc()
+    {
+      if (forward_tmp) { delete forward_tmp; }
+      if (backward_tmp) { delete backward_tmp; }
+    }
+
+    void fill_random(std::vector<transfer_float> &v)
+    {
+      static std::random_device rd;
+      // the good rng
+      static std::mt19937 rng(23ul * comm_rank());
+      // The gaussian distribution
+      static std::normal_distribution<double> n(0., 1.);
+
+      for (auto &x : v) { x = 1e-1 * n(rng); }
+    }
+
+    template <class Base> void apply(Base &base, ColorSpinorField &out, const ColorSpinorField &in)
+    {
+      madwf_ml::transfer_5d_hh<transfer_float, transfer_type>(*forward_tmp, in, device_param, false);
+      base(*backward_tmp, *forward_tmp);
+      madwf_ml::transfer_5d_hh<transfer_float, transfer_type>(out, *backward_tmp, device_param, true);
+
+      blas::axpy(mu, const_cast<ColorSpinorField &>(in), out);
+    }
+
+    template <class Ref, class Base>
+    double cost(const Ref &ref, Base &base, ColorSpinorField &out, const ColorSpinorField &in)
```

Comment:
Is this function called in a hot loop?  If so, the allocation overhead of temporaries may be signficant.

Reply:
I thought if we use memory pool the allocation of these temporaries would be negligible?

Reply:
The actual memory allocation is negligible, but there are other overheads associated with field creation such as zeroing out any padded region (likely not applicable here) and the host-side setup cost of setting the meta data in the class.  Regardless, as a design goal, I think it's desirable to avoid memory allocation / free in hot loops.  The memory pool allocator does lead to an overhead, and in some cases one may need to disable the pool allocator to fit on a desired number of GPUs.

Reply:
Having thought about this a bit more, I think we can just leave the temporary creation.  In the of DWF fermions, this really will be a negligible overhead, and we probably don't need to worry about the pool off case.  If it's an issue in the future, it can always be optimized then.  Marking as resolved.